src/kafka-util/src/client.rs

```diff
@@ -237,20 +242,76 @@ pub struct BrokerRewrite {
 
 /// A client context that supports rewriting broker addresses.
 #[derive(Clone)]
-pub struct BrokerRewritingClientContext<C> {
+pub struct TunnelingClientContext<C> {
     inner: C,
-    rewrites: BTreeMap<BrokerAddr, Arc<dyn Fn() -> BrokerRewrite + Send + Sync>>,
+    rewrites: Arc<Mutex<BTreeMap<BrokerAddr, Arc<dyn Fn() -> BrokerRewrite + Send + Sync>>>>,
+    default_tunnel: Option<SshTunnelConfig>,
+    ssh_tunnel_manager: SshTunnelManager,
+    runtime: Handle,
 }
 
-impl<C> BrokerRewritingClientContext<C> {
+impl<C> TunnelingClientContext<C> {
     /// Constructs a new context that wraps `inner`.
-    pub fn new(inner: C) -> BrokerRewritingClientContext<C> {
-        BrokerRewritingClientContext {
+    pub fn new(
+        inner: C,
+        runtime: Handle,
+        ssh_tunnel_manager: SshTunnelManager,
+    ) -> TunnelingClientContext<C> {
+        TunnelingClientContext {
             inner,
-            rewrites: BTreeMap::new(),
+            rewrites: Arc::new(Mutex::new(BTreeMap::new())),
+            default_tunnel: None,
+            ssh_tunnel_manager,
+            runtime,
         }
     }
 
+    /// Adds the default broker rewrite rule.
+    ///
+    /// Connections to brokers that aren't specified in other rewrites will be rewritten to connect to
+    /// `rewrite_host` and `rewrite_port` instead.
+    pub fn add_default_tunnel(&mut self, tunnel: SshTunnelConfig) {
+        self.default_tunnel = Some(tunnel);
+    }
+
+    /// Adds a broker rewrite rule, using the given tunnel.
+    ///
+    /// `rewrite` is a function that returns a `BrokerRewrite` that specifies
+    /// how to rewrite the address for `broker`.
+    ///
+    /// The function is invoked by librdkafka on every connection attempt to the
+    /// broker. This permits the rewrite to evolve over time, for example, if
+    /// the rewrite is for a tunnel whose address changes if the tunnel fails
+    /// and restarts.
+    pub async fn add_tunnel_rewrite(
```

Comment:
Same justification as above: this only works for SSH tunnels.

```suggestion
    pub async fn add_ssh_tunnel(
```